xdsl/traits.py

```diff
@@ -729,3 +729,60 @@ def get_insn(self, op: Operation) -> str:
         Return the insn representation of the operation for printing.
         """
         raise NotImplementedError()
+
+
+@dataclass(frozen=True)
+class SameOperandsAndResultType(OpTrait):
+    """Constrain the operation to have the same operands and result type."""
+
+    def verify(self, op: Operation) -> None:
+        from xdsl.dialects.builtin import NoneAttr, TensorType
+        from xdsl.utils.type import get_element_type_or_self, have_compatible_shape
+
+        if len(op.results) < 1 or len(op.operands) < 1:
+            raise VerifyException(
+                f"'{op.name}' requires at least one result or operand"
+            )
+
+        res_type0 = get_element_type_or_self(op.result_types[0])
+
+        def get_encoding(maybe_shaped_type: Attribute) -> Attribute:
+            if isinstance(maybe_shaped_type, TensorType):
+                return maybe_shaped_type.encoding
+            return NoneAttr()
```

Comment:
I would recommend moving this to a static method

Reply:
hmm, yeah, static method or put it under `utils/type.py`?

Reply:
Your call, in either case if it's public would be nice to add a unit test for it also